pmemcheck/pmc_main.c

```diff
@@ -1367,6 +1518,7 @@ do_fence(void)
 static void
 do_flush(UWord base, UWord size)
 {
+    Bool array_valid_flush=array_process_flush(base, size);
```

Comment:
this seems to always traverse both the array-stored data and the tree stored data. Is this needed?
Also, there are now two places where the warning is emitted. It would be better to unify this code.

And again, this looks like it could use an abstraction that transparently deals with the underlying data structures.

BTW, do you think it would be possible to have some bloom-filter-like data structure to check where is the info for a given flush?

Reply:
Q1:this seems to always traverse both the array-stored data and the tree-stored data. Is this needed?

A1: Yes, it needs. The array structure is different from a cache and data in array structure are not a part of data in the tree structure. 

Q2: Also, there are now two places where the warning is emitted. It would be better to unify this code.

A2: Yes, you are right. I have done it.

Q3: do you think it would be possible to have some bloom-filter-like data structure to check where is the info for a given flush?

A3: We have to check both structures, so I think it is not necessary to check where is the info for a given flush.